### ISSUE TITLE
Hability to ease rotation when changing bearing

### ISSUE THREAD

**Issue by fnicollet** [hyperknot/Leaflet#30]
Would be really nice to have a rotation transition when changing bearing suddenly, say from 15 to 75.
One way could be to use an easing function inside setBearing, that would manually make the transition. Using the slider in rotate.html works very well, might be the best option.

The other way would be applying CSS transitions. I tried applying some on leaflet-rotate-pane but the map goes a little crazy. I think it is because there is both translate and rotate. And sometimes CSS seem to transition the wrong way (animating the angle by taking the "long route").

**Comment by hyperknot** [comment]
I am afraid that would introduce even more race conditions into animations.
pan and zoom and inertia and maxbounds animations are already hard and
mixing rotate animation into core would allow users to accidentally call
them at the same time.

A better solution I think is to do the animation client side (via RAF for
example).

**Comment by fnicollet** [comment]
when you say "client side", do you mean outside of the library? As in
animate the property (using timeouts or RAF) outside of leaflet/rotate and
ease the calls to setBearing?

On Thu, Jun 30, 2016 at 12:17 PM, Zsolt Ero <notifications@github.com>
wrote:

> I am afraid that would introduce even more race conditions into animations.
> pan and zoom and inertia and maxbounds animations are already hard and
> mixing rotate animation into core would allow users to accidentally call
> them at the same time.
>
> A better solution I think is to do the animation client side (via RAF for
> example).
>
> On Thursday, 30 June 2016, Fabien NICOLLET <notifications@github.com>
> wrote:
>
> > Would be really nice to have a rotation transition when changing bearing
> > suddenly, say from 15 to 75.
> > One way could be to use an easing function inside setBearing, that would
> > manually make the transition. Using the slider in rotate.html works very
> > well, might be the best option.
> >
> > The other way would be applying CSS transitions. I tried applying some on
> > leaflet-rotate-pane but the map goes a little crazy. I think it is
> because
> > there is both translate and rotate. And sometimes CSS seem to transition
> > the wrong way (animating the angle by taking the "long route").
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/hyperknot/Leaflet/issues/30>, or mute the thread
> > <
> https://github.com/notifications/unsubscribe/AAeKj06BvvvYWIuZQJJYAFKkByYGMrl8ks5qQwbRgaJpZM4JBrr6
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperknot/Leaflet/issues/30#issuecomment-229619480>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABLtoS_BQ_mV7aoGWd_UHxTYB8wANeuoks5qQ5fXgaJpZM4JBrr6>
> .
>



-- 
Fabien Nicollet
Tel : +33 (0)6 60 54 72 20

**Comment by hyperknot** [comment]
yes, i mean in user code, not leaflet code. you can then use any easing you
prefer

**Comment by fnicollet** [comment]
I see. I will have to do it in my application, so i will share the code
back once it is done, probably in one of the example pages
Le 30 juin 2016 12:37, "Zsolt Ero" <notifications@github.com> a écrit :

yes, i mean in user code, not leaflet code. you can then use any easing you
prefer


On Thursday, 30 June 2016, Fabien NICOLLET <notifications@github.com> wrote:

> when you say "client side", do you mean outside of the library? As in
> animate the property (using timeouts or RAF) outside of leaflet/rotate and
> ease the calls to setBearing?
>
> On Thu, Jun 30, 2016 at 12:17 PM, Zsolt Ero <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>

> wrote:
>
> > I am afraid that would introduce even more race conditions into
> animations.
> > pan and zoom and inertia and maxbounds animations are already hard and
> > mixing rotate animation into core would allow users to accidentally call
> > them at the same time.
> >
> > A better solution I think is to do the animation client side (via RAF
for
> > example).
> >
> > On Thursday, 30 June 2016, Fabien NICOLLET <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>

> > wrote:
> >
> > > Would be really nice to have a rotation transition when changing
> bearing
> > > suddenly, say from 15 to 75.
> > > One way could be to use an easing function inside setBearing, that
> would
> > > manually make the transition. Using the slider in rotate.html works
> very
> > > well, might be the best option.
> > >
> > > The other way would be applying CSS transitions. I tried applying some
> on
> > > leaflet-rotate-pane but the map goes a little crazy. I think it is
> > because
> > > there is both translate and rotate. And sometimes CSS seem to
> transition
> > > the wrong way (animating the angle by taking the "long route").
> > >
> > > —
> > > You are receiving this because you are subscribed to this thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/hyperknot/Leaflet/issues/30>, or mute the thread
> > > <
> >
>
https://github.com/notifications/unsubscribe/AAeKj06BvvvYWIuZQJJYAFKkByYGMrl8ks5qQwbRgaJpZM4JBrr6
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/hyperknot/Leaflet/issues/30#issuecomment-229619480>,
> > or mute the thread
> > <
>
https://github.com/notifications/unsubscribe/ABLtoS_BQ_mV7aoGWd_UHxTYB8wANeuoks5qQ5fXgaJpZM4JBrr6
> >
> > .

> >
>
>
>
> --
> Fabien Nicollet
> Tel : +33 (0)6 60 54 72 20
>
> —
> You are receiving this because you commented.

> Reply to this email directly, view it on GitHub
> <https://github.com/hyperknot/Leaflet/issues/30#issuecomment-229622024>,
> or mute the thread
> <
https://github.com/notifications/unsubscribe/AAeKj2sknQM-JKOs90Mpqp-PRWTMFTajks5qQ5sCgaJpZM4JBrr6
>

> .
>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/hyperknot/Leaflet/issues/30#issuecomment-229623171>,
or mute
the thread
<https://github.com/notifications/unsubscribe/ABLtobhS3vrQXK7I1lYgqXzEH6HcP_1uks5qQ5xwgaJpZM4JBrr6>
.

**Comment by hyperknot** [comment]
yes I think it's the best thing to do!

**Comment by fnicollet** [comment]
Got it to work, i'll post the way to do here if anyone wants to try it. Code has to be adapted, also could probably be improved (not sure if requesting RAF and cancelling RAF every time is good or not)
The code is mostly taken from http://codepen.io/anon/pen/HGnDF/, which also solve the issue of passing the 0/360 degree bit.
```
GPSController.prototype.rotateRequestId  = -1;
GPSController.prototype.joint = {
        angle: 0,
        targetAngle: 0,
        easing: 0.2,
        velocity: 0,
        maxAccel: 1
    };
GPSController.prototype.clip = function (x, min, max) {
    return x < min ? min : x > max ? max : x
};
GPSController.prototype.doRotateMap = function () {
    if (this.rotateRequestId != -1) {
        window.cancelAnimationFrame(this.rotateRequestId);
        this.rotateRequestId = -1;
    }
    var dtheta = this.joint.targetAngle - this.joint.angle;
    if (Math.abs(dtheta) < 0.01) {
        // rotation is finished
        return;
    }
    if (dtheta > Math.PI){
        this.joint.angle += 2 * Math.PI;
    } else if (dtheta < -Math.PI){
        this.joint.angle -= 2 * Math.PI;
    }

    // then, cap acceleration to eliminate jerkiness

    var targetVel = (this.joint.targetAngle - this.joint.angle) * this.joint.easing;
    this.joint.velocity = this.clip(targetVel,
                          this.joint.velocity - this.joint.maxAccel,
                          this.joint.velocity + this.joint.maxAccel);

    // finally, move with clipped velocity
    this.joint.angle += this.joint.velocity;
    var bearingDeg = this.joint.angle * L.DomUtil.RAD_TO_DEG;
    if (bearingDeg < 0) {
        bearingDeg += 360;
    }
    //console.log("" + bearingDeg + ", " + dtheta + ", " + this._map._bearing);
    var currentBearing = this.map._bearing * L.DomUtil.RAD_TO_DEG;
    if (currentBearing != bearingDeg) {
        this.map.setBearing(bearingDeg);
    }
};
GPSController.prototype.rotateMap = function (radians) {
    if (this.rotateRequestId != -1) {
        window.cancelAnimationFrame(this.rotateRequestId);
        this.rotateRequestId = -1;
    }
    this.joint.targetAngle = radians;
    this.rotateRequestId = requestAnimationFrame(this.doRotateMap.bind(this));
};
```
then just call
```
rotateMap(mapBearing * L.DomUtil.DEG_TO_RAD)
```
with mapBearing being in degrees
